# `--- Day 13: Packet Scanners ---`

In [1]:
import pandas as pd
import numpy as np
from io import StringIO

In [2]:
test_s = '''0: 3
1: 2
4: 4
6: 4'''
test_input = StringIO(test_s)

In [3]:
#fw = pd.read_csv(test_input, sep=':', header=None, names=['depth', 'range'])
fw = pd.read_csv('input.txt', sep=':', header=None, names=['depth', 'range'])

In [4]:
fw.dtypes

depth    int64
range    int64
dtype: object

In [5]:
fw['part1pos'] = fw['depth'] % (2 * fw['range'] - 2)

In [6]:
fw['caught'] = (fw['part1pos'] == 0)
fw['severity'] = fw['caught'] * fw['depth'] * fw['range']
print(f'part1 answer: {fw.severity.sum()}')

part1 answer: 1640


In [7]:
fw['modulus'] = (2 * fw['range'] - 2)
fw = fw.sort_values('modulus')

In [8]:
# %%timeit
# delay = 3960000 # brute force from 0 takes a long time. This starting value chosen after I ran a full brute-force
# while True:
#     fw['part2pos'] = (fw['depth'] + delay) % fw['modulus']
#     #fw['p2caught'] = (fw['part2pos'] == 0)
#     #fw['p2severity'] = fw['p2caught'] * fw['depth'] * fw['range']
#     if ~np.any(fw['part2pos'] == 0):
#         print(f'part 2 answer {delay}')
#         break
#     if delay % 10000 == 0:
#         print(f'{delay}')
#     delay += 1


```3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
426 ms ± 11.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)```

In [10]:
%%timeit
delay = 3960000
while True:
    sol = True
    for i in fw[['depth', 'modulus']].values:
        if (i[0] + delay) % i[1] == 0:
            sol = False
            break
    if sol:
        print(f'part 2 answer {delay}')
        break       
    if delay % 10000 == 0:
        print(f'{delay}')
    delay += 1
    

3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
3960000
part 2 answer 3960702
313 ms ± 18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### pre-meltdown/spectre fix
    3960000
    part 2 answer 3960702
    290 ms ± 9.29 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

    3960000
    part 2 answer 3960702
    295 ms ± 10.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
### post-fix
    3960000
    part 2 answer 3960702
    314 ms ± 17.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

    3960000
    part 2 answer 3960702
    313 ms ± 18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [10]:
fw


depth  range  part1pos  caught  severity  modulus
1       1      2         1   False         0        2
0       0      3         0    True         0        4
2       2      4         2   False         0        6
3       4      4         4   False         0        6
4       6      5         6   False         0        8
6      10      6         0    True        60       10
7      12      6         2   False         0       10
8      14      6         4   False         0       10
9      16      6         6   False         0       10
16     30      8         2   False         0       14
15     28      8         0    True       224       14
21     40      8        12   False         0       14
10     18      8         4   False         0       14
5       8      8         8   False         0       14
11     20      8         6   False         0       14
14     26      9        10   False         0       16
13     24     10         6   False         0       18
25     48     10        12   False         0       18
12     22     12         0    True       264       22
35     70     12         4   False         0       22
31     62     12        18   False         0       22
26     50     12         6   False         0       22
27     52     12         8   False         0       22
20     38     12        16   False         0       22
19     36     12        14   False         0       22
17     32     12        10   False         0       22
18     34     12        12   False         0       22
22     42     12        20   False         0       22
23     44     14        18   False         0       26
38     78     14         0    True      1092       26
37     76     14        24   False         0       26
36     74     14        22   False         0       26
34     68     14        16   False         0       26
32     64     14        12   False         0       26
30     58     14         6   False         0       26
29     56     14         4   False         0       26
28     54     14         2   False         0       26
24     46     14        20   False         0       26
33     66     14        14   False         0       26
42     88     14        10   False         0       26
40     82     17        18   False         0       32
39     80     18        12   False         0       34
41     84     30        26   False         0       58